crates/core/src/metadata_store.rs

```diff
@@ -11,61 +11,150 @@
+    "E0001",
+    help = "Unimplemented Function",
+    description =
+        "This could indicate a misconfigured client or a missing metadata-store role on node."
+);
+
+#[derive(Debug, Clone, thiserror::Error)]
+pub enum InternalError {
+    #[error("{0}")]
+    Message(String),
+    #[error("{0}")]
+    WithCode(String, &'static Code),
+}
+
+impl CodedError for InternalError {
+    fn code(&self) -> Option<&'static Code> {
+        match self {
+            Self::Message(_) => None,
+            Self::WithCode(_, code) => Some(code),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error, CodedError)]
 pub enum ReadError {
+    #[code(unknown)]
     #[error("network error: {0}")]
     Network(GenericError),
     #[error("internal error: {0}")]
-    Internal(String),
+    Internal(
+        #[source]
+        #[code]
+        InternalError,
+    ),
+    #[code(unknown)]
     #[error("codec error: {0}")]
     Codec(GenericError),
+    #[code(unknown)]
     #[error("store error: {0}")]
     Store(GenericError),
 }
 
-#[derive(Debug, thiserror::Error)]
+impl ReadError {
+    pub fn internal(msg: String) -> Self {
+        Self::Internal(InternalError::Message(msg))
+    }
+
+    pub fn internal_with_code(msg: String, code: Option<&'static Code>) -> Self {
+        match code {
+            Some(code) => Self::Internal(InternalError::WithCode(msg, code)),
+            None => Self::internal(msg),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error, CodedError)]
 pub enum WriteError {
+    #[code(unknown)]
     #[error("failed precondition: {0}")]
     FailedPrecondition(String),
+    #[code(unknown)]
     #[error("network error: {0}")]
     Network(GenericError),
     #[error("internal error: {0}")]
-    Internal(String),
+    Internal(
+        #[source]
+        #[code]
+        InternalError,
+    ),
+    #[code(unknown)]
     #[error("codec error: {0}")]
     Codec(GenericError),
+    #[code(unknown)]
     #[error("store error: {0}")]
     Store(GenericError),
 }
 
-#[derive(Debug, thiserror::Error)]
+impl WriteError {
+    pub fn internal(msg: String) -> Self {
+        Self::Internal(InternalError::Message(msg))
+    }
+
+    pub fn internal_with_code(msg: String, code: Option<&'static Code>) -> Self {
+        match code {
+            Some(code) => Self::Internal(InternalError::WithCode(msg, code)),
+            None => Self::internal(msg),
+        }
+    }
+}
+#[derive(Debug, thiserror::Error, CodedError)]
 pub enum ProvisionError {
+    #[code(unknown)]
     #[error("network error: {0}")]
     Network(GenericError),
     #[error("internal error: {0}")]
-    Internal(String),
+    Internal(
+        #[source]
+        #[code]
+        InternalError,
+    ),
+    #[code(unknown)]
     #[error("codec error: {0}")]
     Codec(GenericError),
+    #[code(unknown)]
     #[error("store error: {0}")]
     Store(GenericError),
+    #[code(unknown)]
     #[error("provisioning is not supported: {0}")]
     NotSupported(String),
```

Comment:
I fear that I have voided your work with my latest change to also include Unknown code statuses as retriable errors. Sorry for this.